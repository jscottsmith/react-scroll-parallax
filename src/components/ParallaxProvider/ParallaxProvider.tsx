import React, { Component } from 'react';

import { ParallaxContext } from '../../context/ParallaxContext';
import { ParallaxController, ScrollAxis } from 'parallax-controller';
import { ParallaxProviderProps } from './types';
import { createController } from './helpers';

export class ParallaxProvider extends Component<ParallaxProviderProps, {}> {
  static defaultProps = {
    scrollAxis: ScrollAxis.vertical,
  };

  controller: ParallaxController | null;

  constructor(props: ParallaxProviderProps) {
    super(props);
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
      disabled: props.isDisabled,
    });
  }

  componentDidUpdate(prevProps: ParallaxProviderProps) {
    if (
      prevProps.scrollContainer !== this.props.scrollContainer &&
      this.props.scrollContainer
    ) {
      this.controller?.updateScrollContainer(this.props.scrollContainer);
    }

    if (prevProps.isDisabled !== this.props.isDisabled) {
      if (this.props.isDisabled) {
        this.controller?.disableParallaxController();
      }
      if (!this.props.isDisabled) {
        this.controller?.enableParallaxController();
      }
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
