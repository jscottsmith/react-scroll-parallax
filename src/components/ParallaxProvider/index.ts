export { ParallaxProvider } from './ParallaxProvider';
export * from './types';
