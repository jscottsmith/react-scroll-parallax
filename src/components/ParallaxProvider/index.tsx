import React, { Component } from 'react';

import { ParallaxContext } from '../../context/ParallaxContext';
import { ParallaxController, ScrollAxis } from 'parallax-controller';
import { ParallaxProviderProps } from './types';
import { createController } from './helpers';

export class ParallaxProvider extends Component<ParallaxProviderProps, {}> {
  static defaultProps = {
    scrollAxis: ScrollAxis.vertical,
  };

  controller: ParallaxController | null;

  constructor(props: ParallaxProviderProps) {
    super(props);
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
    });
  }

  componentDidUpdate(prevProps: ParallaxProviderProps) {
    if (prevProps.scrollContainer !== this.props.scrollContainer) {
      // @ts-ignore
      this.controller.updateScrollContainer(this.props.scrollContainer);
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
