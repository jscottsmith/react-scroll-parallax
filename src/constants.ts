export const VERTICAL = 'vertical';
export const HORIZONTAL = 'horizontal';
