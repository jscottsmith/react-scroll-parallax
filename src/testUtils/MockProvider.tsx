import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ParallaxContext } from '../../src/context/ParallaxContext';
import { ParallaxController } from 'parallax-controller';

export class MockProvider extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    controllerMock: PropTypes.object.isRequired,
  };
  controller: ParallaxController;

  constructor(props: { controllerMock: ParallaxController }) {
    super(props);
    this.controller = props.controllerMock;
  }

  componentWillUnmount() {
    if (!this.controller) return;
    this.controller.destroy();
    // @ts-ignore
    this.controller = undefined;
  }

  render() {
    const { children } = this.props;

    return (
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
