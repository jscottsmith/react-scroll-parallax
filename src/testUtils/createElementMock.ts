export function createElementMock(properties = {}, methods = {}) {
  const element = document.createElement('div');
  Object.keys(properties).forEach((key) => {
    Object.defineProperty(element, key, {
      // @ts-ignore
      value: properties[key],
      writable: false,
    });
  });

  Object.keys(methods).forEach((key) => {
    // @ts-ignore
    element[key] = methods[key];
  });

  return element;
}
