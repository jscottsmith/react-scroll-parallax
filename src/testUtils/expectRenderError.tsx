/* eslint-disable */
const React = require('react');
const ReactDOM = require('react-dom');

export default function expectRenderError(
  element: any,
  expectedError: string,
  expectedErrorLength: number = 1
) {
  // Noop error boundary for testing.
  class TestBoundary extends React.Component {
    constructor(props: any) {
      super(props);
      this.state = { didError: false };
    }
    componentDidCatch() {
      this.setState({ didError: true });
    }
    render() {
      return this.state.didError ? null : this.props.children;
    }
  }

  // Record all errors.
  let topLevelErrors: ErrorEvent[] = [];
  function handleTopLevelError(event: ErrorEvent) {
    topLevelErrors.push(event.error);
    // Prevent logging
    event.preventDefault();
  }

  const div = document.createElement('div');
  window.addEventListener('error', handleTopLevelError);
  try {
    // @ts-ignore
    ReactDOM.render(<TestBoundary>{element}</TestBoundary>, div);
  } finally {
    window.removeEventListener('error', handleTopLevelError);
  }

  expect(topLevelErrors.length).toBe(expectedErrorLength);
  expect(topLevelErrors[0].message).toContain(expectedError);
}
