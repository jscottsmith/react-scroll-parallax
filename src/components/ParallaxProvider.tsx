import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ParallaxContext from '../helpers/ParallaxContext';
import ParallaxController from '../classes/ParallaxController';
import { VERTICAL, HORIZONTAL } from '../constants';
import validHTMLElement from '../utils/validHTMLElement';

const createController = (options) => {
  // Don't initialize on the server
  const isServer = typeof window === 'undefined';

  if (!isServer) {
    // Must not be the server so kick it off...
    return ParallaxController.init(options);
  }
  return null;
};

export interface ParallaxProviderProps {
  /**
   * Optionally pass the scroll axis for setting horizontal/vertical scrolling. One of vertical or
   * horizontal
   */
  scrollAxis?: 'vertical' | 'horizontal';
  /**
   * Optionally set the container that has overflow and will contain parallax elements. Defaults
   * to the HTML body
   */
  scrollContainer?: any;
}

export default class ParallaxProvider extends Component<
  ParallaxProviderProps,
  {}
> {
  static defaultProps = {
    scrollAxis: VERTICAL,
  };

  static propTypes = {
    children: PropTypes.node.isRequired,
    scrollAxis: PropTypes.oneOf([VERTICAL, HORIZONTAL]),
    scrollContainer: validHTMLElement,
  };

  constructor(props) {
    super(props);
    // @ts-ignore
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
    });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.scrollContainer !== this.props.scrollContainer) {
      // @ts-ignore
      this.controller.updateScrollContainer(this.props.scrollContainer);
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
