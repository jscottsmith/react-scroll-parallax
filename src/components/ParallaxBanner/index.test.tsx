import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import { ParallaxBanner } from '.';
import { ParallaxProvider } from '../ParallaxProvider';
import { ALL_PARALLAX_PROPS } from '../../testUtils/tests.constants';
import { ParallaxController, ScrollAxis } from 'parallax-controller';
import { MockProvider } from '../../testUtils/MockProvider';

describe('given a <ParallaxBanner> component', () => {
  describe('with all props', () => {
    it('then it will render banners correctly', () => {
      const { asFragment } = render(
        <ParallaxProvider>
          <ParallaxBanner
            className="test-class"
            disabled={false}
            layers={[{ image: 'https://foo.com/bar.jpg', speed: 2 }]}
            style={{ backgroundColor: 'blue', border: '1px solid red' }}
          >
            <div>
              <h1>Foo Bar</h1>
            </div>
          </ParallaxBanner>
        </ParallaxProvider>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe.each(ALL_PARALLAX_PROPS)('when the prop %s is given', (props) => {
    it('then it renders without issue and calls create element with props', () => {
      const controller = ParallaxController.init({
        scrollAxis: ScrollAxis.vertical,
      });
      controller.createElement = jest.fn(controller.createElement);
      function Wrapper(props: PropsWithChildren<{}>) {
        return (
          <MockProvider controllerMock={controller}>
            {props.children}
          </MockProvider>
        );
      }
      const { asFragment } = render(
        <ParallaxBanner layers={[{ ...props }]} />,
        {
          wrapper: Wrapper,
        }
      );
      expect(asFragment()).toMatchSnapshot();
      expect(controller.createElement).toBeCalledWith({
        el: expect.any(HTMLElement),
        props: {
          ...props,
          shouldDisableScalingTranslations: true,
          targetElement: expect.any(HTMLElement),
        },
      });
    });
  });

  describe('when creating layers in the controller', () => {
    it('then it defaults shouldDisableScalingTranslations to true', () => {
      const controller = ParallaxController.init({
        scrollAxis: ScrollAxis.vertical,
      });
      controller.createElement = jest.fn(controller.createElement);
      function Wrapper(props: PropsWithChildren<{}>) {
        return (
          <MockProvider controllerMock={controller}>
            {props.children}
          </MockProvider>
        );
      }
      render(<ParallaxBanner layers={[{ children: <div /> }]} />, {
        wrapper: Wrapper,
      });
      expect(controller.createElement).toBeCalledWith({
        el: expect.any(HTMLElement),
        props: {
          shouldDisableScalingTranslations: true,
          targetElement: expect.any(HTMLElement),
        },
      });
    });
  });
  describe('when children are defined', () => {
    it('then it will render the children', () => {
      const { getByTestId } = render(
        <ParallaxProvider>
          <ParallaxBanner>
            <div data-testid="children" />
          </ParallaxBanner>
        </ParallaxProvider>
      );
      expect(getByTestId('children')).toBeInTheDocument();
    });
  });
  describe('when custom defined layer children are defined', () => {
    it('then it will render each layer child', () => {
      const { getByTestId } = render(
        <ParallaxProvider>
          <ParallaxBanner
            layers={[
              { children: <div data-testid="foo">foo</div> },
              { children: <div data-testid="bar">bar</div> },
            ]}
          />
        </ParallaxProvider>
      );
      expect(getByTestId('foo')).toBeInTheDocument();
      expect(getByTestId('bar')).toBeInTheDocument();
    });
  });

  describe('when the layer expanded option is false', () => {
    it('then it will render without expanded styles', () => {
      const { getByTestId } = render(
        <ParallaxProvider>
          <ParallaxBanner layers={[{ speed: 2, expanded: false }]} />
        </ParallaxProvider>
      );
      expect(getByTestId('layer-0').style.top).toBe('0px');
      expect(getByTestId('layer-0').style.right).toBe('0px');
      expect(getByTestId('layer-0').style.left).toBe('0px');
      expect(getByTestId('layer-0').style.bottom).toBe('0px');
      expect(getByTestId('layer-0').style.position).toBe('absolute');
    });
  });

  describe('when the layer is expanded and', () => {
    describe('when the speed prop is set to a positive number', () => {
      it('then it will render with expanded styles based on speed', () => {
        const { getByTestId } = render(
          <ParallaxProvider>
            <ParallaxBanner layers={[{ speed: 2 }]} />
          </ParallaxProvider>
        );
        expect(getByTestId('layer-0').style.top).toBe('-20px');
        expect(getByTestId('layer-0').style.right).toBe('0px');
        expect(getByTestId('layer-0').style.left).toBe('0px');
        expect(getByTestId('layer-0').style.bottom).toBe('-20px');
        expect(getByTestId('layer-0').style.position).toBe('absolute');
      });
    });
    describe('when the speed prop is set to a negative number', () => {
      it('then it will render with expanded styles based on speed', () => {
        const { getByTestId } = render(
          <ParallaxProvider>
            <ParallaxBanner layers={[{ speed: -4 }]} />
          </ParallaxProvider>
        );
        expect(getByTestId('layer-0').style.top).toBe('-40px');
        expect(getByTestId('layer-0').style.right).toBe('0px');
        expect(getByTestId('layer-0').style.left).toBe('0px');
        expect(getByTestId('layer-0').style.bottom).toBe('-40px');
        expect(getByTestId('layer-0').style.position).toBe('absolute');
      });
    });
    describe('when the translateY prop is set [0px, 10px]', () => {
      it('then it will render with expanded styles based on the translate start end values', () => {
        const { getByTestId } = render(
          <ParallaxProvider>
            <ParallaxBanner layers={[{ translateY: ['0px', '10px'] }]} />
          </ParallaxProvider>
        );
        expect(getByTestId('layer-0').style.top).toBe('-10px');
        expect(getByTestId('layer-0').style.right).toBe('0px');
        expect(getByTestId('layer-0').style.left).toBe('0px');
        expect(getByTestId('layer-0').style.bottom).toBe('0px');
        expect(getByTestId('layer-0').style.position).toBe('absolute');
      });
    });
    describe('when the translateY prop is set [-40px, 30px]', () => {
      it('then it will render with expanded styles based on the translate start end values', () => {
        const { getByTestId } = render(
          <ParallaxProvider>
            <ParallaxBanner layers={[{ translateY: ['-40px', '30px'] }]} />
          </ParallaxProvider>
        );
        expect(getByTestId('layer-0').style.top).toBe('-30px');
        expect(getByTestId('layer-0').style.right).toBe('0px');
        expect(getByTestId('layer-0').style.left).toBe('0px');
        expect(getByTestId('layer-0').style.bottom).toBe('-40px');
        expect(getByTestId('layer-0').style.position).toBe('absolute');
      });
    });
    describe('when the translateY prop is set [0px, 100px]', () => {
      it('then it will render with expanded styles based on the translate start end values', () => {
        const { getByTestId } = render(
          <ParallaxProvider>
            <ParallaxBanner layers={[{ translateY: ['0px', '100px'] }]} />
          </ParallaxProvider>
        );
        expect(getByTestId('layer-0').style.top).toBe('-100px');
        expect(getByTestId('layer-0').style.right).toBe('0px');
        expect(getByTestId('layer-0').style.left).toBe('0px');
        expect(getByTestId('layer-0').style.bottom).toBe('0px');
        expect(getByTestId('layer-0').style.position).toBe('absolute');
      });
    });
  });
  describe('with custom props', () => {
    it('then it will render children', () => {
      const { container, getByTestId } = render(
        <ParallaxProvider>
          <ParallaxBanner
            layers={[
              {
                speed: 2,
                style: {
                  backgroundColor: 'red',
                },
                className: 'my-custom-class',
              },
            ]}
          />
        </ParallaxProvider>
      );
      expect(container.querySelector('.my-custom-class')).toBeInTheDocument();
      expect(getByTestId('layer-0').style.background).toBe('red');
    });
  });
  describe('when custom html props are given', () => {
    it('then it adds them to the returned div', () => {
      const { container, getByTestId } = render(
        <ParallaxProvider>
          <ParallaxBanner
            style={{ background: 'red' }}
            className="my-class"
            id="test-id"
            data-testid="data-test-id"
            data-foo="bar"
            aria-label="Cool"
          />
        </ParallaxProvider>
      );
      expect(getByTestId('data-test-id')).toBeInTheDocument();
      expect(container.querySelector('.my-class')).toBeInTheDocument();
      expect(container.querySelector('#test-id')).toBeInTheDocument();
      expect(getByTestId('data-test-id')).toHaveAttribute('aria-label', 'Cool');
      expect(getByTestId('data-test-id')).toHaveAttribute('data-foo', 'bar');
      expect(getByTestId('data-test-id').style.background).toBe('red');
    });
  });
  describe('with custom props are defined in the layer', () => {
    it('then it adds them to the layer div', () => {
      const { container, getByTestId } = render(
        <ParallaxProvider>
          <ParallaxBanner
            layers={[
              {
                style: { background: 'red' },
                className: 'my-class',
                id: 'test-id',
                // @ts-expect-error
                'data-testid': 'data-test-id',
                'data-foo': 'bar',
                'aria-label': 'Cool',
              },
            ]}
          />
        </ParallaxProvider>
      );
      expect(getByTestId('data-test-id')).toBeInTheDocument();
      expect(container.querySelector('.my-class')).toBeInTheDocument();
      expect(container.querySelector('#test-id')).toBeInTheDocument();
      expect(getByTestId('data-test-id')).toHaveAttribute('aria-label', 'Cool');
      expect(getByTestId('data-test-id')).toHaveAttribute('data-foo', 'bar');
      expect(getByTestId('data-test-id').style.background).toBe('red');
    });
  });
});
