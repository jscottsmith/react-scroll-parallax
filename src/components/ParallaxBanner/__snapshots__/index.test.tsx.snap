// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a <ParallaxBanner> component when the prop { disabled: true } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { easing: 'easeInQuad' } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { easing: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { onChange: [Function: onChange] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { onEnter: [Function: onEnter] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { onExit: [Function: onExit] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { onProgressChange: [Function: onProgressChange] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { opacity: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform,opacity; opacity: 1;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { rootMargin: [Object] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { rotate: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: rotate(1turn);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { rotateX: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: rotateX(1turn);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { rotateY: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: rotateY(1turn);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { rotateZ: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: rotateZ(1turn);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { scale: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: scale(1);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { scaleX: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: scaleX(1);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { scaleY: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: scaleY(1);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { scaleZ: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: scaleZ(1);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { shouldAlwaysCompleteAnimation: true } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { shouldDisableScalingTranslations: true } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { speed: 10 } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: -100px; left: 0px; right: 0px; bottom: -100px; will-change: transform; transform: translateY(-100px);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { startScroll: 0, endScroll: 1000 } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { targetElement: HTMLDivElement {} } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { translateX: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: translateX(100%);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component when the prop { translateY: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      data-testid="layer-0"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform; transform: translateY(100%);"
    />
  </div>
</DocumentFragment>
`;

exports[`given a <ParallaxBanner> component with all props then it will render banners correctly 1`] = `
<DocumentFragment>
  <div
    class="test-class"
    style="position: relative; overflow: hidden; width: 100%; background-color: blue; border: 1px solid red;"
  >
    <div
      data-testid="layer-0"
      style="background-image: url(https://foo.com/bar.jpg); background-position: center; background-size: cover; position: absolute; top: -20px; left: 0px; right: 0px; bottom: -20px; will-change: transform; transform: translateY(-20px);"
    />
    <div>
      <h1>
        Foo Bar
      </h1>
    </div>
  </div>
</DocumentFragment>
`;
